x/ibc/proposal_handler_test.go

```diff
@@ -0,0 +1,162 @@
+
+	// overwrite client FrozenHeight based on the isFrozen flag
+	if isFrozen {
+		clientState.FrozenHeight = 15
+		suite.Require().True(clientState.IsFrozen(), "expected client to be frozen")
+	} else {
+		clientState.FrozenHeight = 0
+		suite.Require().False(clientState.IsFrozen(), "expected client to not be frozen")
+	}
+
+	// overwrite client LatestTimestamp based on the isExpire flag
+	if isExpire {
+		clientState.LatestTimestamp = clientTime
+		suite.Require().True(clientState.Expired(), "expected client to be expired")
+	} else {
+		clientState.LatestTimestamp = time.Now()
+		suite.Require().False(clientState.Expired(), "expected client not to be expired")
+	}
+
+	return clientState
+}
+
+func (suite *ProposalHandlerTestSuite) testStoredClientStatus(exptedHeight uint64, expectedIsFrozen bool, clientID string) {
+	ibcKeeper := *suite.app.IBCKeeper
+	clientkeeper := ibcKeeper.ClientKeeper
+	clientStatus, ok := clientkeeper.GetClientState(suite.ctx, clientID)
+	suite.Require().True(ok)
+	suite.Require().Equal(exptedHeight, clientStatus.GetLatestHeight())
+	suite.Require().Equal(expectedIsFrozen, clientStatus.IsFrozen())
+}
+
+func (suite *ProposalHandlerTestSuite) TestClientUpdateProposalHandler() {
+	testCases := []struct {
+		name        string
+		clientState *ibctmtypes.ClientState
+		expPass     bool
+	}{
+		// tests all combinations of the the following flags:
+		// allowGovernanceOverrideAfterExpire, isExpire
+		// allowGovernanceOverrideAfterMisbehaviour, isFrozen
+		{"test 1", suite.testClientState(false, false, false, false), false},
+		{"test 2", suite.testClientState(false, false, false, true), false},
+		{"test 3", suite.testClientState(false, false, true, false), false},
+		{"test 4", suite.testClientState(false, false, true, true), true},
+		{"test 5", suite.testClientState(false, true, false, false), false},
+		{"test 6", suite.testClientState(false, true, false, true), false},
+		{"test 7", suite.testClientState(false, true, true, false), false},
+		{"test 8", suite.testClientState(false, true, true, true), true},
+		{"test 9", suite.testClientState(true, false, false, false), false},
+		{"test 10", suite.testClientState(true, false, false, true), false},
+		{"test 11", suite.testClientState(true, false, true, false), false},
+		{"test 12", suite.testClientState(true, false, true, true), true},
+		{"test 13", suite.testClientState(true, true, false, false), true},
+		{"test 14", suite.testClientState(true, true, false, true), true},
+		{"test 15", suite.testClientState(true, true, true, false), true},
+		{"test 16", suite.testClientState(true, true, true, true), true},
+	}
+
+	for _, tc := range testCases {
+		suite.Run(tc.name, func() {
+			suite.SetupTest()
+			ibcKeeper := *suite.app.IBCKeeper
+			clientkeeper := ibcKeeper.ClientKeeper
+
+			// create test client
+			_, err := clientkeeper.CreateClient(suite.ctx, clientID, tc.clientState, suite.consensusState)
```

Comment:
There is a check that I want to add to prevent the above situation:

![image](https://user-images.githubusercontent.com/30392990/90700059-18a1b500-e285-11ea-8da5-6032b1035f5f.png)

However, If I do add the check many existing tests will fail. So I am considering to open a new PR for this. But let see, perhaps is not that difficult to fix all the tests.